EduMGTK2020/nodejs2023Q2-service
# Task #8: Implementation Home Library Service: Part 2

Task: https://github.com/AlreadyBored/nodejs-assignments/blob/main/assignments/containerization-database-orm/assignment.md
Screenshot: 
![image](https://github.com/EduMGTK2020/nodejs2023Q2-service/assets/73290069/0efd09a2-0b01-4a5b-95f9-2a70942be21a)

Deploy: it is not required
Done 14.08.2023 / deadline 15.08.2023
Score: 360/ 360

## Steps to get checked:

- Clone repo and checkout to the **part2** branch
- Install dependencies: **npm i**
- Create **.env file** (based on .env.example):  ./.env 
- Build app and db images, run containers in daemon mode and run migrations: **npm run docker:up** - Building images can take several minutes. Be patient - wait for the build process to complete :) 
- For run linter: **npm run lint**
- For tests: **npm run test** for local or **npm run docker:test** in container - **Make sure the app is ready beforehand :)**

## Other commands

- For vulnerabilities scanning: **npm run docker:scan** - Important! You must be logged in to Docker Hub (in **YOUR** profile) to check it out
- For stop and remove containers - **npm run docker:down**
- For full cleanup: **npm run docker:clean** - Be careful - this command will remove ALL images and volumes
- For build images without running - **npm run docker:build**

## A few important notes

- The application container runs in hot-reload mode - you can change the source code in the local ./src folder and the application in the container will restart automatically.
- All migrations are done locally
- If you want to change application or database port settings first stop and remove containers - **npm run docker:down**, make changes in **.env** file and then start again **npm run docker:up**  
- You can see the result (running containers and their output) in Docker Desktop. If you want to see the application running in a local terminal, use **npm run docker:uplocal**, but then you will need a second terminal window to run the tests. And you will need to perform manual migrations (**npm run migration:run**) before launching them.

- If during startup or testing you see error messages such as
```
Error: Connection terminated unexpectedly
Exceeded timeout of 5000 ms for a test
dependency failed to start: container db is unhealthy
```
or something similar, it is most likely that your hardware is not coping with the load. Try full cleanup and restarting everything by stopping all non-test related applications :) 

- You can see the images in the public repo at https://hub.docker.com/u/edumgtk2020 If you want to use them for testing (although this is not required by TOR), uncomment lines 5 and 23 in the docker-compose file - then the images will be downloaded from Docker Hub at startup instead of being build locally. 


## Basic Scope

# Containerization, Docker

- [x] **+20** `Readme.md` has instruction how to run application
- [x]  **+30** `user-defined bridge` is created and configured
- [x]  **+30**  container auto restart after crash
- [x]  **+20** application is restarting upon changes implemented into `src` folder
- [x]  **+30** database files and logs to be stored in volumes instead of container

# Database (PostgreSQL) & ORM

- [x]  **+20** `Users` data is stored in **PostgreSQL** database and `typeorm` / `prisma`  interacts with the database to manipulate data.  
- [x]  **+20** `Artists` data is stored in **PostgreSQL** database and `typeorm` / `prisma`  interacts with the database to manipulate data.
- [x]  **+20** `Albums` data is stored in **PostgreSQL** database and `typeorm` / `prisma`  interacts with the database to manipulate data.
- [x]  **+20** `Tracks` data is stored in **PostgreSQL** database and `typeorm` / `prisma`  interacts with the database to manipulate data.
- [x]  **+20** `Favorites` data is stored in **PostgreSQL** database and `typeorm` / `prisma`  interacts with the database to manipulate data.

## Advanced Scope

# Containerization, Docker

- [x]  **+20** Final size of the Docker image with application is less than 500 MB
- [x]  **+10** Implemented npm script for vulnerabilities scanning (free solution)
- [x]  **+20** Your built image is pushed to DockerHub

# Database & ORM

- [x]  **+30** Migrations are used to create database entities 
- [x]  **+10** Variables used for connection to database to be stored in `.env`
- [x]  **+10** `typeorm` [decorators](https://typeorm.io/#/relations) or `prisma` relations create relations between entities
- [x]  **+30** Local **PostgreSQL** installation is not required for task check, connection is implemented to database stored in `docker` container  (on the basis of the previous task)
